Automattic/wp-calypso
# SPT: immediately render modal with empty thumbnails

#### Changes proposed in this Pull Request

This PR changes slightly the approach in how the thumbnails are dynamically previewed.

_Although it doesn't depend strictly on https://github.com/WordPress/gutenberg/pull/17242, it will improve reduce the previewing time even for all thumbnails and as well as the large preview (100ms)_

#### Overview

The current behavior waits to get the templates parsed before to start to render and mount the components. It produces a delay when the template modal appears. In order to improve the performance and the user perception, I purpose the following more important changes in this PR:

* `Disengage` the parsing process from the react component, starting to process as fast as possible without waiting for the component lifecycle.
* `Do not update the state` with the parsed templates blocks. Update the state[ seems to take considerable time.](https://github.com/Automattic/wp-calypso/pull/35713#issuecomment-536676899)
* Add a `templates-parser.js` which acts as a data store and global event dispatcher. It contains selectors in order to get the desired data.

* `Dispatch onTemplateParse` global event to communicate with the component. The event dispatcher sends all data about templates in the event object argument:

```es6
window.addEventListener( 'onTemplateParse', event => console.log ( event ) );
```

![image](https://user-images.githubusercontent.com/77539/66085323-41bf3700-e547-11e9-8a5d-dbafb3c050b7.png)

The event detail object contains the following fields:

* `blocks`: list of the parsed blocks
* `count`: the number of blocks
* `isEmpty`: `true` is the template doesn't have content.
* `isParsing`: defines if the template is being parsed or not.
* `title`: template title.
* `slug`: template slug.

#### Testing instructions

Apply the patch and confirm that you see the spinners in the thumbnails when the templates are being parsed.

#### Static vs Dynamic


#### Tests


##### Parsing One By One
```
time - blank: 0.005126953125ms
time - about: 46.718017578125ms
time - contact: 19.9072265625ms
time - portfolio: 28.283935546875ms
time - services: 51.375ms
time - team: 37.27197265625ms
time - blog: 29.65673828125ms
time - menu: 51.2978515625ms
time - total: 6304.18798828125ms
```
